File: book2-algorithms-sedgewick/ch2/01_selection_sort.py
import sys
import random
import time

class SelectionSort:

    def sort(self, items) -> None:
        for i in range(len(items)):
            min_index = i
            for j in range(i + 1, len(items)):
                if items[j] < items[min_index]:
                    min_index = j
            self.swap(items, i, min_index)

    def swap(self, items, i, j) -> None:
        items[i], items[j] = items[j], items[i]


def main():
    print('Selection Sort')

    if len(sys.argv) < 2:
        print('Missing number of items.')
        exit(1)
    
    number_items = int(sys.argv[1])
    print('Number of Items: ', number_items)

    print('Generating numbers... ', end='', flush=True)
    numbers = [random.randint(0, 1000000) for _ in range(number_items)]
    print('Done.')

    if number_items <= 100:
        print('Numbers: ', ', '.join((str(n) for n in numbers)))
    
    print('Sorting... ', end='', flush=True)
    selection_sort = SelectionSort()
    t1 = time.time()
    selection_sort.sort(numbers)
    t2 = time.time()
    print('Done.')

    print('Running time: ', (t2 - t1), 'secs.')

    if number_items <= 100:
        print('Numbers: ', ', '.join((str(n) for n in numbers)))

    print('Checking... ', end='', flush=True)
    for i in range(len(numbers) - 1):
        if numbers[i] > numbers[i + 1]:
            print('\nBad ordering: ', numbers[i], ', ', numbers[i + 1])
            exit(1)
    print('Done.')

if __name__ == '__main__':
    main()

File: book2-algorithms-sedgewick/ch2/01_selection_sort.cpp
#include <iostream>
#include <utility>
#include <random>
#include <string>
#include <vector>
#include <chrono>

template<typename T>
class SelectionSort {
public:
    SelectionSort() {};

    void sort(T& arr) {
        size_t len = arr.size();
        for (size_t i = 0; i < len; i++) {
            auto min_index = i;
            for (size_t j = i + 1; j < len; j++)
                if (arr[j] < arr[min_index]) min_index = j;
            swap(arr, i, min_index);
        }
    }
private:
    void swap(T& arr, size_t i, size_t j) {
        auto tmp = arr[i];
        arr[i] = arr[j];
        arr[j] = tmp;
    }
};

int main(int argc, char* argv[]) {
    using std::chrono::high_resolution_clock;
    using std::chrono::duration_cast;
    using std::chrono::milliseconds;

    std::cout << "Selection Sort." << std::endl;

    if (argc <= 1) {
        std::cerr << "Missing number of elements." << std::endl;
        return EXIT_FAILURE;
    }

    size_t numbers_length = std::stoi(argv[1]);
    std::cout << "Number of Elements: " << numbers_length << std::endl;

    std::cout << "Generating data... " << std::flush;
    std::vector<long> numbers;
    std::random_device rd;
    std::mt19937 gen(rd());
    std::uniform_int_distribution <size_t> dist(0, 1000000);
    while (numbers_length > 0) {
        numbers.push_back(dist(gen));
        numbers_length--;
    }
    std::cout << "Done." << std::endl;

    std::cout << "Sorting... " << std::flush;
    auto t1 = high_resolution_clock::now();
    SelectionSort<std::vector<long>> selection_sort;
    selection_sort.sort(numbers);
    auto t2 = high_resolution_clock::now();
    std::cout << "Done." << std::endl;
    auto running_time = duration_cast<milliseconds>(t2 - t1).count();
    std::cout << "Running Time: " << running_time << "ms" << std::endl;

    if (numbers.size() < 100) {
        auto it = numbers.begin();
        std::cout << "Numbers: ";
        while (it != numbers.end()) {
            std::cout << *it << " ";
            it++;
        }
        std::cout << std::endl;
    }

    std::cout << "Checking... " << std::flush;
    for (size_t i = 0; i < numbers.size() - 1; i++) {
        if (numbers[i] > numbers[i + 1]) {
            std::cerr << "Bad sorting: " << numbers[i] << ", " << numbers[i + 1] << std::endl;
            return EXIT_FAILURE;
        }
    }
    std::cout << "Done." << std::endl;

    return EXIT_SUCCESS;
}